rasa/core/actions/action.py

```diff
@@ -1095,7 +1105,9 @@ async def _execute_validation_action(
             event for event in extraction_events if isinstance(event, SlotSet)
         ]
 
-        slot_candidates = "\n".join([e.key for e in slot_events])
+        slot_candidates = "\n".join(
+            [e.key for e in slot_events if isinstance(e, SlotSet)]
```

Comment:
The reason this was set to be `List[Union[Event, SlotSet]]` seems to be to avoid another `mypy` error in L1118:
```
rasa/core/actions/action.py:1118: error: Unsupported operand types for + ("List[Event]" and "List[SlotSet]")  [operator]
```

Reply:
ah yeah, that's the suggestion from [this issue](https://github.com/python/mypy/issues/3933). What does `cast(List[Event], ...)` do at line 1118?

Reply:
That seems to fix the error in L1118, however there's a `return-value` error in L114 when removing `Union`:
```
rasa/core/actions/action.py:1114: error: Incompatible return value type (got "List[SlotSet]", expected "List[Event]")  [return-value]
```

Reply:
this seemed to work ⬇️ WDYT?

<img width="614" alt="Screenshot 2022-03-01 at 17 24 32" src="https://user-images.githubusercontent.com/1931615/156207786-21c81ce6-f75a-4bbb-91c6-945137612a15.png">

Reply:
it's explicitly casting, but not changing what is done at runtime (i.e. not checking multiple times for `isinstance()`)

Reply:
Ahaa, this works indeed! The problem that I didn't realise was that I wasn't casting the first return to `List[Event]`. All good now, will push these changes.